daml-lf/spec/daml-lf-1.rst

```diff
@@ -3151,68 +3134,144 @@ in which keys were inserted into the map (insertion-order).
 * ``GENMAP_INSERT : ∀ α. ∀ β.  α → β → 'GenMap' α β → 'GenMap' α β``
 
   Inserts a new key and value in the map. If the key is already
-  present in the map, the associated value is replaced with the
-  supplied value, otherwise the new key/value entry is appended at the
-  ends of the map.
-
-  This raises an error if the key is not a valid map key. Keys are
-  compared according to ``~ᵥ``.
+  present according the builtin function ``EQUAL``, the associated
+  value is replaced with the supplied value, otherwise the key/value
+  is inserted in order according the builtin function ``LESS`` applied
+  on keys. This raises a runtime error if it tries to compare
+  incomparable values.
```

Comment:
One question that the spec should answer: does it raise a runtime error when inserting a value that is incomparable in an empty map? (I think it should raise an error, for the sake of consistency.)

Reply:
It does not. 
It raise an exception exactly once `LESS` or `EQUAL` does. 

We agreed with @hurryabit and @hurryabit that is the only semantics that is reasonable. Otherwise you will have to always traverse all the keys to ensure there is no function inside.

Reply:
Ok, if @hurryabit and @hurryabit can confirm ;-) 

The reason I don't like this is that you are creating the potential for latent runtime errors when you perform inserts -- i.e. creating a runtime error that will only be found on a subsequent lookup, instead of when the real error occurs. This is particularly bad for e.g. people using genmaps during testing, or choice validation.

When adding a key to the map, It should be easy to compare a key to itself using EQUAL, to make sure it doesn't introduce latent runtime errors.

Reply:
In the proposed semantics, you are only saving 1 comparison, but you are introducing latent runtime errors, so I don't think it's worth it.

Reply:
It depends on what kind of values you have. If you have ADT-like trees, being able to short circuit on comparisons can make a pretty large difference. Given that we prevent this runtime error in DAML, I’m not sure that missing out on better performance here in favor of “nicer semantics” for DAML-LF is worth it. That said, I don’t feel particularly strongly about this.

Reply:
In fact, @associahedron I like your proposal.
The advantage I see, is that we can switch the implementation to hashmap, if we realize the implementation is too slow (we will sorted the map only for serialization). 
 
That is too say in each case (insertion, deletion and lookup) we check the value is equal to itself. Hence we have a nice definition of "comparable" value.

Reply:
@cocreature If you use nested trees as keys, maps are going to be slow regardless. I'm also not confident that this case will never ever occur at runtime in DAML.

@remyhaemmerle-da 👍

Reply:
Alright, then let’s error out for now.